---
title: 'Report generated on `r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    theme: flatly
    include:
        before_body: Logo.svg
editor_options: 
  chunk_output_type: console
params:  
  version: "No Found"
  github: "No Found"
  fastqc_r1: "No Found"
  fastqc_r2: "No Found"
  fastqc_trim_r1: "No Found"
  fastqc_trim_r2: "No Found"
  novoassembly_path: "No Found"
  chromosome_path: "No Found"
  plasmid_path: "No Found"
  phage_path: "No Found"
  novoassembly_path_QC: "No Found"
  prokka_denovo_path: "No Found"
  prokka_chr_path: "No Found"
  prokka_plasmid_path: "No Found"
  pharokka_path: "No Found"
  plasmid_class: "No Found"
  phage_class: "No Found"
  snippy_path: "No Found"
  gtdb_path: 'No Found'
  minimap2_path: "No Found"
  assembly2gene_table: "No Found"
  assembly2gene_aligments: "No Found"
  assembly2gene_peptides: "No Found"
  stats_path: 'No Found'
  fastqc_version: 'version'
  trimmomatic_version: 'version'
  unicycler_version: 'version'
  snippy_version: 'version'
  prokka_version: 'version'
  pharokka_version: 'version'
  plasclass_version: 'version'
  checkv_version: 'version'
  checkm_version: 'version'
  gtdb_version: 'version'
  minimap2_version: 'version'

 
---




```{r setup, include=FALSE}

all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)


knitr::opts_chunk$set(warning=FALSE, message = FALSE, echo=FALSE,fig.width =8,fig.height = 6)

```
 


```{r results='asis'}
cat( "<div class='textright'> source code " , params$github, "</div>   \n")
cat( "<div class='textright'> ProkGenomics ", params$version,"</div>   \n")
```


*** 
# **Results**

#### **Reads quality**

```{r results='asis'}
cat( "- **Quality control for forward read:** ", params$fastqc_r1,"   \n")
cat( "- **Quality control for reverse read:** ", params$fastqc_r2,"   \n")
cat( "- **Quality control for forward read after trimming:** ", params$fastqc_trim_r1,"   \n")
cat( "- **Quality control for reverse read  after trimming:** ", params$fastqc_trim_r2,"   \n")

```


#### **Extrachromosomal prediction tables**

```{r results='asis'}
cat( "- **Plasmid classifications:** ", params$plasmid_class,"   \n")
cat( "- **Phage classifications:** ", params$phage_class,"   \n")

```


#### **Assemblies**

- **De novo assembly**

```{r results='asis'}

cat( "  - **fasta:** ", params$novoassembly_path,"   \n")
cat( "  - **QC:** ", params$novoassembly_path_QC,"   \n")
cat( "  - **Annotations** ", params$prokka_denovo_path,"   \n")
```

- **Chromosome**

```{r results='asis'}

cat( "  - **fasta:** ",  params$chromosome_path,"   \n")
cat( "  - **Annotations** ", params$prokka_chr_path,"   \n")
cat( "  - **Classification** ", params$gtdb_path,"   \n")
```

- **Plasmids**

```{r results='asis'}

cat( "  - **fasta:** ",  params$plasmid_path,"   \n")
cat( "  - **Annotations** ", params$prokka_plasmid_path,"   \n")
```

- **Phage**

```{r results='asis'}

cat( "  - **fasta:** ",  params$phage_path,"   \n")
cat( "  - **Annotations** ", params$pharokka_path,"   \n")
```

#### **Comparative genomics**

```{r results='asis'}
cat( "- **Single Nucleotide Variants (SNV):** ", params$snippy_path,"   \n")
cat( "- **Mapping:** ", params$minimap2_path,"   \n")
cat( "- **Coverage Stats:** ", params$stats_path,"   \n")



```

#### **Characterization of genes of interest**

```{r results='asis'}
cat( "- **Genes summary table:** ", params$assembly2gene_table,"   \n")
cat( "- **Genes aligments:** ", params$assembly2gene_aligments,"   \n")
cat( "- **Peptides:** ", params$assembly2gene_peptides,"   \n")

```

***

# **Methods Summary**

```{r results='asis'}

cat("|Tool|Version|Description|Paper\n")
cat("| :---|:---- |:---|:---|\n")
cat( "|[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)|",params$fastqc_version,"|Sequence quality controls|Andrews S. (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc|\n")
cat( "|[Trimmomatic](https://github.com/timflutre/trimmomatic)|",params$trimmomatic_version,"|Trim primer adaptor from reads|Bolger AM, Lohse M, Usadel B. Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics. 2014 Aug 1;30(15):2114-20. doi: 10.1093/bioinformatics/btu170. Epub 2014 Apr 1. PMID: 24695404; PMCID: PMC4103590. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4103590/|\n")
cat( "|[Unicycler](https://github.com/rrwick/Unicycler)|",params$unicycler_version,"|De novo assembly|Wick RR, Judd LM, Gorrie CL, Holt KE (2017) Unicycler: Resolving bacterial genome assemblies from short and long sequencing reads. PLOS Computational Biology 13(6): e1005595. https://doi.org/10.1371/journal.pcbi.1005595|\n")
cat( "|[Plasclass](https://github.com/Shamir-Lab/PlasClass)|",params$plasclass_version,"|Plasmids prediction|Pellow D, Mizrahi I, Shamir R (2020) PlasClass improves plasmid sequence classification. PLOS Computational Biology 16(4): e1007781. https://doi.org/10.1371/journal.pcbi.1007781|\n")
cat( "|[CheckV](https://bitbucket.org/berkeleylab/checkv/src/master/)|",params$checkv_version,"|Phage prediction|Nayfach, S., Camargo, A.P., Schulz, F. et al. CheckV assesses the quality and completeness of metagenome-assembled viral genomes. Nat Biotechnol 39, 578–585 (2021). https://doi.org/10.1038/s41587-020-00774-7|\n")
cat( "|[Prokka](https://github.com/tseemann/prokka)|",params$prokka_version,"|Prokaryotic gene annotation|Torsten Seemann, Prokka: rapid prokaryotic genome annotation, Bioinformatics, Volume 30, Issue 14, July 2014, Pages 2068–2069, https://doi.org/10.1093/bioinformatics/btu153|\n")
cat( "|[Pharokka](https://github.com/gbouras13/pharokka)|",params$pharokka_version,"|Phage gene annotation|Bouras G, Nepal R, Houtak G, Psaltis AJ, Wormald PJ, Vreugde S. Pharokka: a fast scalable bacteriophage annotation tool. Bioinformatics. 2023 Jan 1;39(1):btac776. doi: 10.1093/bioinformatics/btac776. PMID: 36453861; PMCID: PMC9805569. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9805569/|\n")
cat( "|[CheckM](https://github.com/Ecogenomics/CheckM)|",params$checkm_version,"|Assembly quality controls|Parks DH, Imelfort M, Skennerton CT, Hugenholtz P, Tyson GW. CheckM: assessing the quality of microbial genomes recovered from isolates, single cells, and metagenomes. Genome Res. 2015 Jul;25(7):1043-55. doi: 10.1101/gr.186072.114. Epub 2015 May 14. PMID: 25977477; PMCID: PMC4484387. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4484387/|\n")
cat( "|[Snippy](https://github.com/tseemann/snippy)|",params$snippy_version,"|Single Nucleotide Variant (SNV) detection|Seemann T (2015). snippy: fast bacterial variant calling from NGS reads. https://github.com/tseemann/snippy|\n")
cat( "|[Minimap2](https://github.com/lh3/minimap2)|",params$minimap2_version,"|Sequence alignment to reference genome|Heng Li, New strategies to improve minimap2 alignment accuracy, Bioinformatics, Volume 37, Issue 23, December 2021, Pages 4572–4574, https://doi.org/10.1093/bioinformatics/btab705|\n")
cat( "|[GTDB](https://github.com/Ecogenomics/GTDBTk)|",params$gtdb_version,"|Taxonomic classifications |Donovan H Parks, Maria Chuvochina, Christian Rinke, Aaron J Mussig, Pierre-Alain Chaumeil, Philip Hugenholtz, GTDB: an ongoing census of bacterial and archaeal diversity through a phylogenetically consistent, rank normalized and complete genome-based taxonomy, Nucleic Acids Research, Volume 50, Issue D1, 7 January 2022, Pages D785–D794, https://doi.org/10.1093/nar/gkab776|\n")







```



<style type="text/css">

h1.title {
  font-size: 16px;
  color: DarkGray;
  text-align: right;
  
}

.textright {
  font-size: 16px;
  color: DarkGray;
  text-align: right;

}
</style>
